### PR TITLE
add support for oc_service for labels, externalIPs

### PR DESCRIPTION
**Pull request by richm**
Add support for `labels` and `externalIPs` in Services
@ewolinetz @jcantrill PTAL

### CODE REVIEW

**Comment by richm** [comment]
fixes https://github.com/openshift/openshift-ansible/issues/4090

**Comment by richm** [comment]
@kwoodson

**Comment by jcantrill** [comment]
LGTM

**Comment by sdodson** [comment]
What's the minimum version of OCP required to use this?

**Comment by richm** [comment]
@sdodson not sure - label and externalIPs support has been in since at least 1.4, probably earlier

**Comment by sdodson** [comment]
Cool, 3.4 or larger is fine.

**Comment by ashcrow** [comment]
Will review after changes made.

**Comment by richm** [comment]
@kwoodson I can't figure out how registry and router process the labels parameter.  There is one reference to labels in ansible/oc_adm_registry.py, in ansible/oc_adm_router.py, class/oc_adm_registry.py, in class/oc_adm_router.py - and in none of those places is the labels argument actually used.  Also, why list?  For example:
```
$ oc get pod router-1-z4knc -o yaml
...
  labels:
    deployment: router-1
    deploymentconfig: router
    router: router
```
This looks like a `dict` to me.  If it is a `list`, what is the format in the yaml representation?

**Comment by richm** [comment]
```
# oadm router --help|grep labels
      --labels='router=<name>': A set of labels to uniquely identify the router and its components.
# oadm router -o yaml
    labels:
      router: router
```

**Comment by richm** [comment]
I think I figured out how to do this, and with testing too. PTAL.

**Comment by kwoodson** [comment]
@richm,  apologies, I thought this was Eric's PR.  Comments below.

I'm not sure we've tested labels with the router/registry creation.  I just don't think that's something we needed to modify.  Now with that being said, we should support it and the idea behind the params being a list is that it would get sent to the command line like this:
```--labels=key=val,key1=val1```

You are correct that there is no processing of labels.  This is why my comment above.  There should be a step before it gets sent to the ServiceConfig.  Should be building a the structure like this:
```python
options['labels'] = (['{}={}'.format(key, val) for key,val in params['labels'].items()]).join(',')
```

Does this make sense?

Thanks for picking this up and adding support.

**Comment by richm** [comment]
@kwoodson What about `--label` or `--labels` being a multi-valued flag e.g. `--label key1=value1 --label key2=value2 ...`

**Comment by ashcrow** [comment]
Travis noted:
```
./roles/lib_openshift/src/test/unit/test_oc_adm_registry.py:221:47: E231 missing whitespace after ':'
```

File: roles/lib_openshift/library/oc_adm_registry.py [hunk 1]
```diff
@@ -2049,6 +2056,7 @@ class Service(Yedit):
     cluster_ip = "spec.clusterIP"
     selector_path = 'spec.selector'
     kind = 'Service'
+    external_ips = "spec.externalIPs"
```
Comment (ashcrow) [inline roles/lib_openshift/library/oc_adm_registry.py]: Not related to this PR directly, but at some point these attributes should probably be renamed to follow a pattern such as ``cluster_ip_key``, etc..

File: roles/lib_openshift/library/oc_adm_registry.py [hunk 2]
```diff
@@ -2110,6 +2118,53 @@ def add_portal_ip(self, pip):
         '''add cluster ip'''
         self.put(Service.portal_ip, pip)
 
+    def get_external_ips(self):
+        ''' get a list of external_ips '''
+        return self.get(Service.external_ips) or []
```
Comment (ashcrow) [inline roles/lib_openshift/library/oc_adm_registry.py]: Is there a reason to not use this instances attribute: ``self.external_ips``? Or are you following the pre set pattern in the file?

Comment (richm) [inline roles/lib_openshift/library/oc_adm_registry.py]: I'm just following the pre set pattern in the file.

**Comment by richm** [comment]
@kwoodson 
> There should be a step before it gets sent to the ServiceConfig. Should be building a the structure like this:
```
options['labels'] = (['{}={}'.format(key, val) for key,val in params['labels'].items()]).join(',')
```

Where?  I'm looking at the code for oc_adm_registry and router.  It looks like the function `to_option_list` is responsible for transforming the internal structure into the proper format for the cli.  This function would take `labels: {"foo":"bar", "baz":"biff"}`.  AFAICT, the best place to do this would be in `to_option_list`, to make it convert the `labels` value `dict` into a comma delimited list as you have above.

**Comment by richm** [comment]
@kwoodson @ashcrow ptal

**Comment by ashcrow** [comment]
Looks like there are some formatting issues https://travis-ci.org/openshift/openshift-ansible/jobs/236488145#L184

**Comment by richm** [comment]
@kwoodson @ashcrow any clues about what I should do about coverage?  Is that a merge blocker?  The problem is that I changed the `to_option_list()` code which `generate.py` includes in every single python file - a lot more code changes + very little coverage change == small percentage decrease in coverage.

**Comment by ashcrow** [comment]
@richm coverage is not a blocker. I'll kick the tests off again now.

**Comment by ashcrow** [comment]
aos-ci-test

**Comment by richm** [comment]
> error: aos-ci-jenkins/OS_3.6_containerized for a1cfec9 (logs)

flake?

**Comment by sdodson** [comment]
aos-ci-test

**Comment by sdodson** [comment]
Yes, you got hit by container images being newer/older than rpm repo because a build was in flight.

**Comment by sdodson** [comment]
aos-ci-test

**Comment by sdodson** [comment]
we'll manually merge tomorrow morning if this fails again

**Comment by richm** [comment]
> we'll manually merge tomorrow morning if this fails again

Thanks!

**Comment by richm** [comment]
@sdodson ready to merge, or do you want to try to aos-ci-test again?

**Comment by ashcrow** [comment]
@richm it should be merged now 🎆

**Comment by richm** [comment]
@ashcrow yep, sorry, I jumped the gun

**Pull request merged**